text/0000-cortex-m-targets.md

```diff
@@ -0,0 +1,595 @@
+  targets that do static linking (`i686` and `x86_64` variants).
+- `target-endian`. (For conditional compilation). The Cortex-M architecture can be either little or
+  big endian. However, for microcontrollers, endianness is chosen at manufacture time and, as far as
+  the author knows, most (or all?) devices are little endian.
+- `target-pointer-width`. (For conditional compilation). All the existing Cortex-M processors are
+  32-bit processors.
+- `cpu`. (Enables CPU specific optimizations). The value pretty much matches the target name.
+- `features`. (Used to control the generation of FPU instructions). Only needed on the Cortex-M4 and
+  Cortex-M7 targets.
+- `llvm-target`. (Controls code generation). Meaning of each field of the triple:
+  - `thumbv*m`. `thumb` indicates that the targets will use the Thumb instruction set instead of the
+    (older) ARM instruction set ("These processors support the Thumb instruction set only. -
+    [ARM]"). The `v*m` bit indicates the version/variant of the ARM architecture: `ARMv6-M`,
+    `ARMv7-M` or `ARMv7E-M`.
+  - `none`. Indicates the lack of an operating system.
+  - `eabi`. Indicates that the target uses the soft-float calling convention instead of the
+    hard-float calling convention (more details about this below).
+
+[ARM]: http://infocenter.arm.com/help/index.jsp?topic=/com.arm.doc.set.cortexm/index.html
+
+### Calling convention
+
+Targets that have a FPU can choose between two calling conventions (CC): *soft-float* or
+*hard-float*. Objects compiled with different CCs have different ABIs and can't always be linked
+together. Although the *hard-float* CC generates faster code around function calls that need to pass
+floating point arguments, this RFC proposes using the soft-float CC for these targets. The reason is
+that this ensures maximum interoperability with pre-compiled C libraries. For example, even if a C
+library has been compiled for the Cortex-M3, which doesn't have a FPU, it can still be linked with
+objects compiled for the `cortex-m4f` target because they also use the soft-float ABI.
+
+## The `rust-core` component
+
+The build system will learn to package a new `rust-core` component that will look like this:
+
+```
+$ tree .
+.
+└── lib
+     └── rustlib
+        └── cortex-m3
+            │── liballoc-$HASH.rlib
+            │── libcollections-$HASH.rlib
+            │── libcompiler-rt.a
+            │── libcore-$HASH.rlib
+            │── librand-$HASH.rlib
+            └── librustc_unicode-$HASH.rlib
+```
+
+This new component will be similar to the existing `rust-std` component. The only difference is that
+the `rust-core` component will ship with a smaller set of standard crates; only the ones that are
+*freestanding*. A freestanding crate is a crate that doesn't depend on an underlying OS, kernel or
+runtime. Currently, this is the full list of freestanding crates:
+
+- `alloc`
```

Comment:
Hmm, the crates listed here compile just fine for Cortex-M processors. (See [Xargo](https://github.com/japaric/xargo#overview)). I haven't (yet) tried to write a program (for these targets) that uses `alloc` + `collections`. But, `liballoc` doesn't seem to contain any undefined reference to a `libc` symbol (AFAICT):

```
$ arm-none-eabi-nm -u liballoc-bdc8710db91bbbe1.rlib

alloc-bdc8710db91bbbe1.0.o:
         U __aeabi_memcpy
         U __aeabi_unwind_cpp_pr0
         U __rust_allocate
         U __rust_deallocate
         U __rust_reallocate
         U __rust_reallocate_inplace
         U __rust_usable_size
         U _ZN4core9panicking5panic17h322e6888d03463ddE
```